extensions/wikia/Email/EmailController.class.php

```diff
@@ -0,0 +1,279 @@
+<?php
+
+namespace Email;
+
+abstract class EmailController extends \WikiaController {
+	const DEFAULT_TEMPLATE_ENGINE = \WikiaResponse::TEMPLATE_ENGINE_MUSTACHE;
+
+	/** @var \User The user associated with the current request */
+	protected $currentUser;
+
+	/** @var \User The user this email will be sent to */
+	protected $targetUser;
+
+	protected $hasErrorResponse = false;
+
+	/** @var bool Whether or not to actually send an email */
+	protected $test;
+
+	/**
+	 * Since the children of this class are located in the 'Controller' directory, the default
+	 * location for the template directory would be 'Controller/templates'.  Redefine it to be
+	 * at the root of this extension instead.
+	 *
+	 * @return string
+	 */
+	public static function getTemplateDir() {
+		return dirname( __FILE__ ) . '/templates';
+	}
+
+	public function init() {
+		try {
+			$this->assertCanAccessController();
+
+			$this->currentUser = $this->findUserFromRequest( 'currentUser', $this->wg->User );
+			$this->targetUser = $this->findUserFromRequest( 'targetUser', $this->wg->User );
+			$this->test = $this->getRequest()->getVal( 'test', false );
+
+			$this->initEmail();
+		} catch ( ControllerException $e ) {
+			return $this->setErrorResponse( $e );
+		}
+	}
+
+	/**
+	 * Make sure to only allow authorized use of this extension.
+	 *
+	 * @throws \Email\Fatal
+	 */
+	public function assertCanAccessController() {
+		if ( $this->wg->User->isStaff() ) {
+			return;
+		}
+
+		if ( $this->request->isInternal() ) {
+			return;
+		}
+
+		throw new Fatal( wfMessage( 'emailext-error-restricted-controller' )->escaped() );
+	}
+
+	/**
+	 * Allow child classes to set some initial values.  Added here so its always defined
+	 * but is kept blank so does not need to be called via parent::initEmail()
+	 */
+	public function initEmail() {
+		// Can be overridden in child class
+	}
+
+	/**
+	 * This is the main entry point for the email extension.  The template set for this
+	 * method is used for testing only, to preview the email that will be sent.
+	 *
+	 * @template Email_preview
+	 *
+	 * @return bool
+	 *
+	 * @throws \MWException
+	 */
+	public function handle() {
+		// If something previously has thrown an error (likely 'init') don't continue
+		if ( $this->hasErrorResponse ) {
+			return true;
```

Comment:
Well, its a little bit of the tail wagging the dog, but it was because I wanted to return on L98 directly from the ```setErrorResponse call```.  PHPStorm complains if ```setErrorResponse``` returns ```null``` and I try to return that, and if I return a value there, the other two returns have to return something.  I can just make L98 call ```setErrorResponse``` and then put the return on the line below.  Thanks for making bloat the code! ;)